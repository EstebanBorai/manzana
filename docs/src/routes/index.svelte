<script lang="ts">
	import { newForm } from 'manzana';
	import * as Yup from 'yup';

	const { initialValues, handleChange, values, errors } = newForm({
		initialValues: {
			name: 'Esteban',
			email: 'hello@estebanborai.com'
		},
		validateOnChange: true,
		validationSchema: Yup.object({
			name: Yup.string()
				.required('You must specify your name.')
				.min(3, 'Your name must have at least 3 character of length'),
			email: Yup.string()
				.required('You must specify an email.')
				.email('This is not a valid email!'),
			age: Yup.number()
				.required('You must specify your age.')
				.moreThan(18, 'You must be 18 years old to signup!')
		})
	});
</script>

<div id="container">
	<section>
		<h1>Hello {$values.name || ''}</h1>
		<h3>I will send you an email to: {$values.email || ''}</h3>

		<form action="" method="get" class="form-example">
			<div class="form-example">
				<label for="name">Enter your name: </label>
				<input type="text" name="name" id="name" required bind:value={$values.name} />
				<span>{$errors.name}</span>
			</div>
			<div class="form-example">
				<label for="email">Enter your email: </label>
				<input
					type="email"
					name="email"
					id="email"
					required
					value={$values.email}
					on:change={handleChange}
				/>
				<span>{$errors.email}</span>
			</div>
			<div class="form-example">
				<label for="age">Enter your age: </label>
				<input type="number" name="age" id="age" required bind:value={$values.age} />
				<span>{$errors.age}</span>
			</div>
			<div class="form-example">
				<input type="submit" value="Subscribe!" />
			</div>
		</form>
	</section>
	<section>
		<code>
			Initial Values:
			<br />
			<pre>
{JSON.stringify(initialValues, undefined, 2)}
</pre>
		</code>
		<br /><br />
		<code>
			Values:
			<br />
			<pre>
{JSON.stringify($values, undefined, 2)}
</pre>
		</code>
		<code>
			Errors:
			<br />
			<pre>
{JSON.stringify($errors, undefined, 2)}
</pre>
		</code>
	</section>
</div>

<style>
	#container {
		display: flex;
	}

	#container section {
		width: 50%;
	}

	#container section:last-of-type {
		background-color: #fafafa;
		box-sizing: border-box;
		padding: 1rem;
	}

	#container section:last-of-type pre {
		color: #aa0000;
	}
</style>
