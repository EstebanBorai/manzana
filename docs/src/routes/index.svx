<script>
  import YupValidation from '$lib/components/example/YupValidation.svelte';
</script>

Manazana brings a Form constructor which provides total control on your
forms. All of the form behavior is handled in the `<script>` side
of your component, this means that Manzana doesn't bring any components to
wrap or build your form on _top of_, instead an instance of a Form object is
created which methods to control and retrieve form data.

## Installation

Manzana is available as a JavaScript package, published on NPM.
You can install it as usually using your favorite package manager:

### NPM

```bash
npm install manzana
```

### Yarn

```bash
yarn add manzana
```

### PNPM

```bash
pnpm add manzana
```

> You don't have to install a `@types/manzana` package, given that manzana is
written on TypeScript and already ports it's own type definitions.

## Usage

The following is a basic form built using manzana to handle the submit event and
store of input values.

```svelte
<script lang="ts">
  import { newForm } from 'manzana';

  const { handleSubmit, values } = newForm({
    initialValues: {
      message: '',
    },
    onSubmit(values) {
      console.log(values);
    }
  });
</script>

<form on:submit={handleSubmit}>
  <input
    type="text"
    name="message"
    bind:value={$values.message}
  />
  <button type="submit">Send</button>
</form>
```

### Validation

Manzana provides form validation support through Yup. When creating a new form
with `newForm` function, you can provide a Yup validation schema which will be
used to validate the form.

```svelte
<script lang="ts">
  import { newForm } from 'manzana';
  import * as Yup from 'yup';

  let valuesSubmitted = null;

  const { errors, handleSubmit, isSubmitting, isValidating, values } = newForm({
    initialValues: {
      sender: '',
      receiver: '',
      message: ''
    },
    validationSchema: Yup.object({
      sender: Yup.string().email().required(),
      receiver: Yup.string().email().required(),
      message: Yup.string().required()
    }),
    onSubmit(values) {
      valuesSubmitted = values;
    }
  });
</script>
```

The following form uses the `script` above, the [full source code is available
on GitHub][1].

<YupValidation />

[1]: https://github.com/EstebanBorai/manzana/blob/main/docs/src/lib/components/example/YupValidation.svelte