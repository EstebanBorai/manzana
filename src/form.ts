/* eslint-disable @typescript-eslint/no-explicit-any */

import { get, writable } from 'svelte/store';

import type { Writable } from 'svelte/store';
import type { SchemaLike } from 'yup/lib/types';
import type { ValidationError } from 'yup';

export type Values = Record<string, any>;

export type FormInstance<T = Values> = {
  /**
   * Form errors.
   *
   * A writable store that holds form validation errors.
   */
  errors: Writable<Partial<T>>;
  /**
   * Form's fields initial values.
   *
   * This object is required to initialize a new form, it's used to track
   * changed values and to initialize the form values.
   */
  initialValues: T;
  /**
   * Event handler for the input's `change` event.
   */
  handleChange(event: Event): void;
  /**
   * Event handler for the input's `input` event.
   *
   * The `handleInput` function comes handy when building `input` components
   * which must support dynamic `type` values, given that using dynamic type
   * attribute along with two-way binding is not legal on Svelte.
   *
   * ```
   * Error: 'type' attribute cannot be dynamic if input uses two-way binding.
   * ```
   *
   * Read the [Rich Harris's answer on StackOverflow][1] for more details on
   * this topic.
   *
   * [1]: https://stackoverflow.com/a/57393751/9888500
   */
  handleInput(event: Event): void;
  /**
   * Imperatively sets the error message for the field with the name provided.
   *
   * The `message` param could be skipped to clear the error message for the
   * field in question.
   */
  setFieldError(name: string, message?: string): void;
  /**
   * Imperatively sets the value for the field with the name provided.
   */
  setFieldValue(name: string, value: any, shouldValidateField?: boolean): void;
  /**
   * Validates a single field using the provided `validationSchema`
   * asynchronously.
   */
  validateField(field: string): Promise<void>;
  /**
   * Validates a single field using the provided `validationSchema`
   * synchronously.
   */
  validateFieldSync(field: string): void;
  /**
   * Current form values.
   *
   * A writable store which holds form's field values. These are initialized
   * by the `initialValues` provided to the form configuration object.
   *
   * Useful to take advantage of the two-way data binding when providing it
   * to the `<input>` HTML element.
   *
   * ## Example
   *
   * ```html
   * <script>
   * import { newForm } from 'manzana';
   *
   * const { values } = newForm({
   *   initialValues: {
   *     name: '',
   *     last: '',
   *   }
   * });
   * </script>
   *
   * <form>
   *   <input name="name" bind:value={$values.name} />
   *   <input name="name" bind:value={$values.name} />
   * </form>
   * ```
   *
   * ## Notes
   *
   * Is not recommended to write directly to this object, instead you must
   * use setFieldValue` instead.
   */
  values: Writable<T>;
};

export type FormConfig<T = Values> = {
  /**
   * Form's fields initial values.
   *
   * This object is required to initialize a new form, it's used to track
   * changed values and to initialize the form values.
   */
  initialValues: T;
  /**
   * Wether to validate form fields whenever `handleChange` is executed.
   */
  validateOnChange?: boolean;
  /**
   * Wether to validate form fields whenever `handleInput` is executed.
   */
  validateOnInput?: boolean;
  /**
   * A [Yup][1] schema used to validate form values.
   *
   * [1]: https://github.com/jquense/yup
   */
  validationSchema?: SchemaLike;
};

/**
 * Retrieves a HTML Input element instance value based on the input's type.
 *
 * Given a type, `getInputValue` will retrieve the value as follows:
 *
 * - `number` or `range`: Retrieve the `number` equivalent using `+` sign.
 * - As fallback retrieves the value _as is_.
 */
export const getInputValue = (inputElement: HTMLInputElement): any => {
  const type = inputElement.type;

  if (type.match(/^(number|range)$/)) {
    return +inputElement.value;
  }

  return inputElement.value;
};

export function newForm<T = Values>(config: FormConfig<T>): FormInstance<T> {
  if (typeof config === 'undefined') {
    throw new TypeError(
      'You must provide a config to "newForm". Expected "config" to be an object, received "undefined" instead.',
    );
  }

  if (!config.initialValues) {
    throw new TypeError(
      'You must specify "initialValues" with an object to "FormConfig".',
    );
  }

  const initialValues = JSON.parse(JSON.stringify(config.initialValues));
  const errors = writable(
    Object.fromEntries(
      Object.keys(initialValues).map((field) => [field, undefined]),
    ) as Partial<T>,
  );
  const values = writable({
    ...initialValues,
  });

  const setFieldError = (field: string, message?: string): void => {
    errors.update((currentState) => ({
      ...currentState,
      [field]: message,
    }));
  };

  /**
   * Updates the `errors` store with an instance of Yup's `ValidationError`.
   */
  const setYupValidationErrors = (error: ValidationError): void => {
    if ('path' in error && Array.isArray(error?.errors)) {
      setFieldError(error.path, error.errors[0]);
      return;
    }

    console.error('Missing "path" and "errors" array.');
  };

  const validateField = async (field: string): Promise<void> => {
    try {
      const currentFormValues = get(values);

      await config.validationSchema.validateAt(field, currentFormValues);
      setFieldError(field, null);
    } catch (error) {
      setYupValidationErrors(error);
    }
  };

  const validateFieldSync = (field: string): void => {
    try {
      const currentFormValues = get(values);

      config.validationSchema.validateSyncAt(field, currentFormValues);
      errors.update((currentState) => ({
        ...currentState,
        [field]: undefined,
      }));
    } catch (error) {
      setYupValidationErrors(error);
    }
  };

  const setFieldValue = (
    name: string,
    value: any,
    shouldValidateField?: boolean,
  ): void => {
    values.update((currentValues) => ({
      ...currentValues,
      [name]: value,
    }));

    if (shouldValidateField && config.validationSchema) {
      validateFieldSync(name);
    }
  };

  const handleChange = (event: Event): void => {
    const target = event.target as HTMLInputElement;
    const name = target.name;
    const value = getInputValue(target);

    return setFieldValue(name, value, config.validateOnChange);
  };

  const handleInput = (event: Event): void => {
    const target = event.target as HTMLInputElement;
    const name = target.name;
    const value = getInputValue(target);

    return setFieldValue(name, value, config.validateOnInput);
  };

  return {
    errors,
    handleChange,
    handleInput,
    initialValues,
    setFieldError,
    setFieldValue,
    values,
    validateField,
    validateFieldSync,
  };
}
