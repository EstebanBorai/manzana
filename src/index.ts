/* eslint-disable @typescript-eslint/no-explicit-any */

import { derived, get, writable } from 'svelte/store';
import { isSchema, ValidationError } from 'yup';

import type { Readable, Writable } from 'svelte/store';
import type { SchemaLike } from 'yup/lib/types';

export type Values = Record<string, any>;

export type FormErrors = Record<string, string | undefined>;

export type FormInstance<T = Values> = {
  /**
   * Form errors.
   *
   * A writable store that holds form validation errors.
   */
  errors: Readable<FormErrors>;

  /**
   * Form's fields initial values.
   *
   * This object is required to initialize a new form, it's used to track
   * changed values and to initialize the form values.
   */
  initialValues: T;

  /**
   * A readable store which holds a boolean `true` if the form submition is
   * being executed.
   */
  isSubmitting: Readable<boolean>;

  /**
   * A readable store which holds a boolean `true` if the form submition is
   * under validation stage.
   */
  isValidating: Readable<boolean>;

  /**
   * Event handler for the input's `change` event.
   */
  handleChange(event: Event): void;

  /**
   * Event handler for the input's `input` event.
   *
   * The `handleInput` function comes handy when building `input` components
   * which must support dynamic `type` values, given that using dynamic type
   * attribute along with two-way binding is not legal on Svelte.
   *
   * ```
   * Error: 'type' attribute cannot be dynamic if input uses two-way binding.
   * ```
   *
   * Read the [Rich Harris's answer on StackOverflow][1] for more details on
   * this topic.
   *
   * [1]: https://stackoverflow.com/a/57393751/9888500
   */
  handleInput(event: Event): void;

  /**
   * Event handler for the form's submit event.
   *
   * Handles the submit event for the `form` element and prevents the
   * default behavior (`event.preventDefault`).
   *
   * Sets the `isSubmitting` store to `true` which can be used as a sentinel
   * value for a loading UI. If the `validationSchema` is available in the
   * `FormConfig`, then the `isValidating` store value will be `true` as well.
   */
  handleSubmit(event: Event): void;

  /**
   * Imperatively sets the error message for the field with the name provided.
   *
   * The `message` param could be skipped to clear the error message for the
   * field in question.
   */
  setFieldError(field: keyof T, message?: string): void;

  /**
   * Imperatively sets the value for the field with the name provided.
   */
  setFieldValue(
    field: keyof T,
    value: any,
    shouldValidateField?: boolean,
  ): void;

  /**
   * Validates a single field using the provided `validationSchema`
   * asynchronously.
   */
  validateField(field: keyof T): Promise<void>;

  /**
   * Validates a single field using the provided `validationSchema`
   * synchronously.
   */
  validateFieldSync(field: keyof T): void;

  /**
   * Current form values.
   *
   * A writable store which holds form's field values. These are initialized
   * by the `initialValues` provided to the form configuration object.
   *
   * Useful to take advantage of the two-way data binding when providing it
   * to the `<input>` HTML element.
   *
   * ## Example
   *
   * ```html
   * <script>
   * import { newForm } from 'manzana';
   *
   * const { values } = newForm({
   *   initialValues: {
   *     name: '',
   *     last: '',
   *   }
   * });
   * </script>
   *
   * <form>
   *   <input name="name" bind:value={$values.name} />
   *   <input name="name" bind:value={$values.name} />
   * </form>
   * ```
   *
   * ## Notes
   *
   * Is not recommended to write directly to this object, instead you must
   * use setFieldValue` instead.
   */
  values: Writable<T>;
};

export type FormConfig<T = Values> = {
  /**
   * Form's fields initial values.
   *
   * This object is required to initialize a new form, it's used to track
   * changed values and to initialize the form values.
   */
  initialValues: T;

  /**
   * Callback to execute when `handleSubmit` is invoked.
   *
   * Form values are provided to this callback as the first argument and
   * `helpers` to update form state.
   */
  onSubmit(values: T, helpers: any): Promise<void> | void;

  /**
   * Wether to validate form fields whenever `handleChange` is executed.
   */
  validateOnChange?: boolean;

  /**
   * Wether to validate form fields whenever `handleInput` is executed.
   */
  validateOnInput?: boolean;

  /**
   * A [Yup][1] schema used to validate form values.
   *
   * [1]: https://github.com/jquense/yup
   */
  validationSchema?: SchemaLike;
};

/**
 * Retrieves a HTML Input element instance value based on the input's type.
 *
 * Given a type, `getInputValue` will retrieve the value as follows:
 *
 * - `number` or `range`: Retrieve the `number` equivalent using `+` sign.
 * - As fallback retrieves the value _as is_.
 */
export const getInputValue = (inputElement: HTMLInputElement): any => {
  const type = inputElement.type;

  if (type.match(/^(number|range)$/)) {
    return +inputElement.value;
  }

  return inputElement.value;
};

/**
 * Creates a new form instance with an `errors` and `values` store.
 *
 * # Example
 *
 * ```html
 * <script lang="ts">
 * import { newForm } from 'manzana';
 *
 * import { productService } from '$lib/services/product';
 *
 * const { handleSubmit, errors, values, setFieldValue } = newForm<{
 *   name: string;
 *   logo: File;
 * }>({
 *   initialValues: {
 *     name: '',
 *     logo: null
 *   },
 *   onSubmit: async (values) => {
 *     await productService.createBrand(values.name, values.logo);
 *   }
 * });
 * </script>
 *
 * <form on:submit={handleSubmit}>
 *   <div>
 *     <Input
 *       name="name"
 *       label="Brand Name"
 *       placeholder="Apple Inc."
 *       bind:value={$values.name}
 *       error={$errors.name}
 *     />
 *     <FileDropzone
 *       error={$errors.logo}
 *       onChange={(files) => setFieldValue('logo', files[0])}
 *     />
 *   </div>
 *   <div class="text-center pt-4">
 *     <Button type="submit">Crear</Button>
 *   </div>
 * </form>
 * ```
 */
export type NewFormFn = <T = Values>(config: FormConfig<T>) => FormInstance<T>;

export const newForm: NewFormFn = <T>(
  config: FormConfig<T>,
): FormInstance<T> => {
  if (typeof config === 'undefined') {
    throw new TypeError(
      'You must provide a config to "newForm". Expected "config" to be an object, received "undefined" instead.',
    );
  }

  if (!config.initialValues) {
    throw new TypeError(
      'You must specify "initialValues" with an object to "FormConfig".',
    );
  }

  const initialValues = JSON.parse(JSON.stringify(config.initialValues));
  const __isSubmitting = writable(false);
  const __isValidating = writable(false);
  const __errors = writable(
    Object.fromEntries(
      Object.keys(initialValues).map((field) => [field, undefined]),
    ) as FormErrors,
  );
  const values = writable({
    ...initialValues,
  });
  const setFieldError = (field: keyof T, message?: string): void => {
    __errors.update((currentState) => ({
      ...currentState,
      [field]: message,
    }));
  };

  /**
   * Updates the `errors` store with an instance of Yup's `ValidationError`.
   */
  const setYupValidationErrors = (error: ValidationError): void => {
    if ('path' in error && Array.isArray(error?.errors)) {
      setFieldError(error.path as keyof T, error.errors[0]);
      return;
    }

    console.error('Missing "path" and "errors" array.');
  };

  const validateField = async (field: keyof T): Promise<void> => {
    try {
      const currentFormValues = get(values);

      await config.validationSchema.validateAt(
        field as string,
        currentFormValues,
      );
      setFieldError(field, null);
    } catch (error) {
      setYupValidationErrors(error);
    }
  };

  const validateFieldSync = (field: keyof T): void => {
    try {
      const currentFormValues = get(values);

      config.validationSchema.validateSyncAt(
        field as string,
        currentFormValues,
      );
      __errors.update((currentState) => ({
        ...currentState,
        [field]: undefined,
      }));
    } catch (error) {
      setYupValidationErrors(error);
    }
  };

  const setFieldValue = (
    field: keyof T,
    value: any,
    shouldValidateField?: boolean,
  ): void => {
    values.update((currentValues) => ({
      ...currentValues,
      [field]: value,
    }));

    if (shouldValidateField && config.validationSchema) {
      validateFieldSync(field);
    }
  };

  const handleChange = (event: Event): void => {
    const target = event.target as HTMLInputElement;
    const name = target.name as keyof T;
    const value = getInputValue(target);

    return setFieldValue(name, value, config.validateOnChange);
  };

  const handleInput = (event: Event): void => {
    const target = event.target as HTMLInputElement;
    const name = target.name;
    const value = getInputValue(target);

    return setFieldValue(name as keyof T, value, config.validateOnInput);
  };

  const handleSubmit = async (event: Event): Promise<void> => {
    if (config?.onSubmit && typeof config.onSubmit === 'function') {
      if (event?.preventDefault && typeof event.preventDefault === 'function') {
        event.preventDefault();
      }

      if (
        event?.stopPropagation &&
        typeof event.stopPropagation === 'function'
      ) {
        event.stopPropagation();
      }

      const currentValues = get(values);

      __isSubmitting.set(true);

      if (isSchema(config.validationSchema)) {
        try {
          __isValidating.set(true);

          config.validationSchema.validate(currentValues, {
            abortEarly: false,
          });
        } catch (error) {
          if (error instanceof ValidationError) {
            if ('inner' in error) {
              const validationErrors = error.inner.reduce(
                (acc: FormErrors, { message, path }) => {
                  return {
                    ...acc,
                    [path]: message,
                  };
                },
                {},
              );

              __errors.set(validationErrors);
              return;
            }

            console.error('An unhandled error ocurred validating the form.');
            console.error(error);
          }
        } finally {
          __isValidating.set(false);
        }
      }

      await config.onSubmit(currentValues, {
        setFieldError,
      });
    }
  };

  return {
    errors: derived(__errors, (errors) => errors),
    handleChange,
    handleInput,
    handleSubmit,
    initialValues,
    isSubmitting: derived(__isSubmitting, (isSubmitting) => isSubmitting),
    isValidating: derived(__isValidating, (isValidating) => isValidating),
    setFieldError,
    setFieldValue,
    values,
    validateField,
    validateFieldSync,
  };
};
