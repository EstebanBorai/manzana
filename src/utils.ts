/**
 * Clones an object with an optional default value to override every field's
 * value.
 *
 *
 * @param obj - Object to clone
 * @param defaultValue - Value to override every object's value
 * @returns Clone from `obj`
 */
export function clone<T extends object, U>(
  obj: T,
  defaultValue: U = undefined,
): T | Record<keyof T, U> {
  return Object.fromEntries(
    Object.keys(obj).map((field) => {
      if (typeof defaultValue !== 'undefined') {
        return [field, defaultValue];
      }

      if (Array.isArray(obj[field])) {
        return [field, [...obj[field]]];
      }

      if (
        typeof obj[field] === 'object' &&
        !Array.isArray(obj[field]) &&
        obj[field] !== null
      ) {
        return [field, { ...obj[field] }];
      }

      return [field, obj[field]];
    }),
  ) as T;
}

export function diff(base: object, other: object) {
  return Object.keys(base).reduce((res, key) => {
    if (!other.hasOwnProperty(key)) {
      // Object `other` is missing a field from `base` object
      res.push(key);
      return res;
    }

    const baseValue = base[key];
    const otherValue = other[key];

    if (Array.isArray(baseValue) && Array.isArray(otherValue)) {
      if (baseValue === otherValue) {
        // Same reference and thus same array
        return res;
      }

      if (baseValue.length !== otherValue.length) {
        res.push(key);
        return res;
      }

      const includesTheSameElements = baseValue.every((el) =>
        otherValue.includes(el),
      );

      if (!includesTheSameElements) {
        res.push(key);
        return res;
      }
    }

    if (
      typeof baseValue === 'object' &&
      !Array.isArray(baseValue) &&
      typeof otherValue === 'object' &&
      !Array.isArray(otherValue) &&
      baseValue !== null &&
      otherValue !== null
    ) {
      const differences = diff(baseValue, otherValue);

      console.log(differences);

      if (differences.length) {
        res.push(key);
        return res;
      }
    }

    if (baseValue !== otherValue) {
      res.push(key);
      return res;
    }

    return res;
  }, []);
}
