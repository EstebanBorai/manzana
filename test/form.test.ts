import { describe, expect, it } from 'vitest';

import { newForm } from '../src';

import type { FormConfig, Values } from '../src/form';

describe('Form: initialValues', () => {
  it('Creates a new form with initial values', () => {
    const initialValues = {
      name: 'Esteban',
      last: 'Borai',
      email: 'esteban@mail.com',
    };
    const form = newForm<{ name: string; last: string; email: string }>({
      initialValues,
    });

    let values: Values = {};

    form.values.subscribe((state) => (values = state));

    expect(values.name).toStrictEqual(initialValues.name);
    expect(values.last).toStrictEqual(initialValues.last);
    expect(values.email).toStrictEqual(initialValues.email);
  });

  it('Uses a copy of `initialValues` instead of relying on the object reference', () => {
    const initialValues = {
      name: 'Esteban',
      last: 'Borai',
      email: 'esteban@mail.com',
    };
    const form = newForm<{ name: string; last: string; email: string }>({
      initialValues,
    });

    initialValues.name = 'John';
    initialValues.last = 'Appleseed';
    initialValues.email = 'john@mail.com';

    expect(form.initialValues.name).toStrictEqual('Esteban');
    expect(form.initialValues.last).toStrictEqual('Borai');
    expect(form.initialValues.email).toStrictEqual('esteban@mail.com');
  });

  it('Throws a "TypeError" if no "initialValues" are provided to the configuration.', () => {
    expect(() => newForm({} as FormConfig)).toThrowErrorMatchingSnapshot();
  });
});

describe('Form: $values', () => {
  it('`form.$values` from `newForm` is a store with a subscribe function', () => {
    const form = newForm({ initialValues: {} });

    expect(form.values.subscribe).toBeDefined();
  });

  it('Holds "initialValues" when subscribing for the first time', () => {
    const initialValues = { name: 'Lorem', last: 'Ipsum' };
    const form = newForm<typeof initialValues>({ initialValues });

    let values = {} as typeof initialValues;

    form.values.subscribe((state) => (values = state));

    expect(values.name).toStrictEqual('Lorem');
    expect(values.last).toStrictEqual('Ipsum');
  });
});
