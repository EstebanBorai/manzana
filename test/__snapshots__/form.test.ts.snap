// Vitest Snapshot v1

exports[`Form: initialValues > Throws a "TypeError" if no "initialValues" are provided to the configuration. 1`] = `"You must specify \\"initialValues\\" with an object to \\"FormConfig\\"."`;
